Adding pip demo and reworking slides for Python packaging (#116)
Simulation-Software-Engineering/Lecture-Material: 03_building_and_packaging/pip_demo.md

# `pip`: Python Packaging Tool Demo Notes

**Note**: Maybe run everything in a fresh Ubuntu Docker container. Launch such a container using `docker run --rm -it ubuntu:jammy`. The other option is to work in a Python virtual environment. An environment can be created as follows

```bash
python -m venv env-name
source env-name/bin/activate
```

## 0. Getting `pip`

If Python has been installed using a system package manager like `apt`, it will not come with `pip`. `pip`can be installed in several ways

```bash
python -m ensurepip --upgrade
```

or

```bash
apt install pip
```

> - If installed with Python, the Python version is used to determine which version of `pip` is installed. So Python 3.8 will install `pip3.8`.

Reviewer: pip has an independent versioning

```
$ pip3 --version
pip 22.2.2 from /home/uekermbn/.local/lib/python3.8/site-packages/pip (python 3.8)
```

Author: Ah indeed this is confusing! The version I point to is the Python version to which a particular `pip` is compatible with, but of course `pip` itself has independent versioning. I will clarify this in the same point.

# `pip`: Python Packaging Tool (Notes)

> **Note**: Maybe run everything in a fresh Docker container. I built the container `jaustar/spack-package-tutorial-base` locally for myself. It contains installed/downloaded Spack, that was not set up yet, i.e., one needs to run `spack compiler find` and `spack external find`. `clingo` needs to be downloaded first as well.

Reviewer: yes, i would recommend to use a completely fresh Docker container or a Python `venv`.

Author: I am still hesitant to use Docker here, because the students have not learnt Docker yet and I do not want to start using things which have not been taught. I will suggest `venv` here

## 1. Installing packages with `pip`

Installing a package

```bash
pip install micro-manager-precice
```

> 

Reviewer: In the very beginning you could explain how to get `pip` in the first place and that `pip` has versions as well.
And maybe that there might be a difference between `pip` and `pip3`.

Install a specific version of a package

```bash
pip install micro-manager-precice==0.0.2
```

Uninstalling a package

```bash
pip uninstall micro-manager-precice
```

- By default `pip` tries to install packages in the system directory, for example `/usr/local/lib/python<version>`. This is problematic as it has no user isolation and it requires root privileges.
- With the `--user` flag, packages are installed in the `home` directory, typically at `/home/user/.local/lib/python<version>/site-packages/`. This is important when one is not working with virtual environments.

Installing a package from a folder

```bash
pip install --user .
```

- A `build` folder is created and the package files are copied into it.

> - But now `pip3 uninstall package-name` does not work as `pip` looks in the home path. Removal of package can be done by deleting the `build` folder or using manually using `setup.py`.

Reviewer: ```suggestion
- But now `pip uninstall package-name` does not work as `pip` looks in the home path. Removal of package can be done by deleting the `build` folder or manually using `setup.py`.
```

> **NOTE**: To successfully install `micro-manager-precice`, the dependencies `libopenmpi-dev`, `pkg-config`and `precice` which can be installed in the following way

Reviewer: ```suggestion
**NOTE**: To successfully install `micro-manager-precice`, the dependencies `libopenmpi-dev`, `pkg-config`, and `precice` are required, which can be installed via Aptitude.
```

Reviewer: Let's be consistent with **Note** vs **NOTE** (see above), please add to the style guide.

```bash
apt install libopenmpi-dev pkg-config
apt install wget
wget https://github.com/precice/precice/releases/download/v2.5.0/libprecice2_2.5.0_jammy.deb
apt install ./libprecice2_2.5.0_jammy.deb

> ```

Reviewer: I am not so sure if preCICE is the best example here. Maybe Nutils could be easier?
`lipopenmpi-dev` should also be part of `libprecice2_2.5.0_jammy.deb`, right?

## 2. Getting information of currently installed packages

```bash
pip list

> ```

Reviewer: Will this search both user and system installation?

Author: This searches the system installation as far as I know, let me confirm.

or for a specific package

```bash
pip show micro-manager-precice
```

## 3. Installing dependencies of a package

- Using `requirements.txt` is a standard way to handle dependency installation.

```bash
pip install -r requirements.txt

> ```

Reviewer: Which example do you plan to use here?
This is only true if you install from a local code, right? Otherwise `pip` already handles the dependencies, no?

Author: I will remove this because in hindsight, using a `requirements.txt` is only done in the development phase and most deployed packages do not ship with this file.

## 4. Running `pip` as a Python module

```bash
python -m pip install package-name
```

## 5. Getting information of installed packages

Get package information on the terminal using

```bash
pip show package-name

> ```

Reviewer: You have this above already
